### DISCUSSION TITLE
enhance automatic filter lookups

### DISCUSSION THREAD

**Discussion by gecBurton** [carltongibson/django-filter#1474] [Q&A] [answered]
The [docs](https://github.com/carltongibson/django-filter/blob/3635b2b67c110627e74404603330583df0000a44/django_filters/filterset.py#L297) specify that the filters can be automatically built from either a list of fields or `__all__` but the [code](
https://github.com/carltongibson/django-filter/blob/3635b2b67c110627e74404603330583df0000a44/django_filters/filterset.py#L297) appears to only produce the `DEFAULT_LOOKUP_EXPR`.

Is it not possible to fetch all possible lookups like:

```python
{field.name, list(field.class_lookups)) for field in model._meta.fields}
```
or have i not understood something?

many thanks!

**Comment by carltongibson** [comment, marked as answer]
See [Generating filters with Meta.fields](https://django-filter.readthedocs.io/en/stable/guide/usage.html#generating-filters-with-meta-fields) in the usage guide. 

If you want to customise the lookups you can pass a dict (or declare the equivalent filters manually): 

```python
class ProductFilter(django_filters.FilterSet):
    class Meta:
        model = Product
        fields = {
            'price': ['lt', 'gt'],
            'release_date': ['exact', 'year__gt'],
        }
```

> Is it not possible to fetch all possible lookups like...

How you generate that dict is up to you. 

See also [`LookupChoiceFilter`](https://django-filter.readthedocs.io/en/stable/ref/filters.html#lookupchoicefilter)

**Reply by gecBurton** [reply]
> How you generate that dict is up to you.

could django-filter automatically generate some sensible defaults for me? 

for example:

```python
class Product(Model):
    name = CharField(max_length=256)
    price = FloatField()
    first_on_sale = DateField()

print({field.name: list(field.class_lookups) for field in Product._meta.fields})
```
would give:
```json
{
  "id": [
    "gte",
    "lt",
    "contained_by"
  ],
  "name": [
    "exact",
    "iexact",
    "gt",
    "gte",
    "lt",
    "lte",
    "in",
    "contains",
    "icontains",
    "startswith",
    "istartswith",
    "endswith",
    "iendswith",
    "range",
    "isnull",
    "regex",
    "iregex"
  ],
  "price": [
    "contained_by"
  ],
  "first_on_sale": [
    "year",
    "month",
    "day",
    "week_day",
    "iso_week_day",
    "week",
    "iso_year",
    "quarter",
    "contained_by"
  ]
}
```
?

**Reply by carltongibson** [reply]
That’s not something I’d add to the package. 

In general you want to be very conservative about which DB options you expose via URLs. Automatically allowing all registered lookups is more than you’ll need.

**Reply by gecBurton** [reply]
Thank you for the explanation @carltongibson :+1:
